# Lambda School Data Science Module 123

## Introduction to Bayesian Inference




## Assignment - Code it up!

We used pure math to apply Bayes Theorem to drug tests. Now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up.

Specific goals/targets:

### 1) Write a function 

`def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate):` 

You should only truly need these two values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?

### 2) Explore `scipy.stats.bayes_mvs`  
Read its documentation, and experiment with it on data you've tested in other ways earlier this week.
 - Create a visualization comparing the results of a Bayesian approach to a traditional/frequentist approach. (with a large sample size they should look close to identical, however, take this opportunity to practice visualizing condfidence intervals in general. The following are some potential ways that you could visualize confidence intervals on your graph:
  - [Matplotlib Error Bars](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.errorbar.html)
  - [Seaborn barplot with error bars](https://seaborn.pydata.org/generated/seaborn.barplot.html)
  - [Vertical ines to show bounds of confidence interval](https://www.simplypsychology.org/confidence-interval.jpg)
  - [Confidence Intervals on Box Plots](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.boxplot.html)

### 3) In your own words, summarize the difference between Bayesian and Frequentist statistics

If you're unsure where to start, check out [this blog post of Bayes theorem with Python](https://dataconomy.com/2015/02/introduction-to-bayes-theorem-with-python/).



## Bayes Theorem

$$ P(A|B) = \frac{P(B|A) \times P(A)}{P(B)}$$

The probability of A given B is equal to the product of probability of B given A multiplied by the probability of A all over the probability of B

if $P(A)$ is the probability that someone is a drug user, and $P(B)$ is the probability of getting a positive result, we can figure out the probability that a drug user will have a positive result.

### Drug Tests
* Given:
  * Probability of a drug user testing positive = 1
  * Probability of a non-drug user testing positive = .01
  * Probability of someone being a drug user = 1/200


In [0]:
# TODO - code!
def bayes_testing(
    prior, 
    sensitivity, 
    specificity, 
    num_tests = 1):
    p_b_given_a = sensitivity 
    numerator = p_b_given_a * prior

    # Add probability of user + with non-user+

    denominator = numerator + ((specificity * (1-prior)))
    result = numerator / denominator
    num_tests -= 1
    if num_tests > 0:
        return bayes_testing(result, sensitivity, specificity, num_tests)
    else:
        return result


In [2]:
bayes_testing(1/200,1,.01, 1)

0.33444816053511706

#### Result
The probability of a randomly selected individual with a positive result is a drug user is $33.4\%$

### Drunk Driving
You should only truly need these two values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?


In [3]:
prior = 1/1000
sensitivity = 1
specificity = .08

bayes_testing(prior, sensitivity, specificity, 1)

0.012357884330202669

In [4]:
bayes_testing(prior, sensitivity, specificity, 2)

0.13525210993291495

In [0]:
def bayes_loop(a, b, c):

    probability = a
    for i in range(1,1000):
        probability = bayes_testing(probability, b, c)
        print(probability)
        if probability >= .95:
            
            return i

In [6]:
bayes_loop(prior, sensitivity, specificity)

0.012357884330202669
0.13525210993291495
0.6615996951348605
0.9606895076105054


4

Real world Breathalyzers. [Citation](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4281544/)

> The digital multiuse breathalyser had a sensitivity of 89.5% (95% CI 75.9% to 95.8%) and a specificity of 64.1% (95% CI 56.6% to 71.0%)

> 20% of fatal accidents involve a drunk driver
[Source](https://crashstats.nhtsa.dot.gov/Api/Public/ViewPublication/812630)


In [8]:
real_world_bayes = bayes_testing(.2, .895, .641)

print(f'The real world probablity of a driver actually being drunk given a positive test is: {real_world_bayes*100}%')

The real world probablity of a driver actually being drunk given a positive test is: 25.874530211043655%


In [9]:
real_world_99 = bayes_loop(.2, .895, .641)

print(f'In order to have a confidence >95%, you would need to test {real_world_99} times')

0.25874530211043656
0.32767802977154475
0.40494283271265413
0.4872234530248339
0.5702023902785309
0.64941555487407
0.7211685627204596
0.7831400397848495
0.8344989562720116
0.8756262557672732
0.907664196591361
0.9320894111790656
0.9504065483325859
In order to have a confidence >95%, you would need to test 13 times


## Resources

- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)

## Stretch Goals:

- Go back and study the content from Modules 1 & 2 to make sure that you're really comfortable with them.
- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!